transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -389,13 +393,47 @@ protected boolean offerChannel(Channel channel) {
 
     @Override
     public void close() {
+        closeIdleChannels(nextEventLoop());
+    }
+
+    void closeIdleChannels(EventExecutor executor) {
+        Promise<Void> result = executor.newPromise();
+        closeIdleChannels(result);
+        if (!executor.inEventLoop()) {
+            result.syncUninterruptibly();
+        }
+    }
+
+    private void closeIdleChannels(final Promise<Void> result) {
+        final Set<Channel> channelsToClose = Collections.newSetFromMap(new ConcurrentHashMap<Channel, Boolean>());
         for (;;) {
             Channel channel = pollChannel();
             if (channel == null) {
                 break;
             }
-            // Just ignore any errors that are reported back from close().
-            channel.close().awaitUninterruptibly();
+            channelsToClose.add(channel);
+        }
+        if (channelsToClose.isEmpty()) {
+            // no idle channels in the pool - nothing to close
+            result.trySuccess(null);
+        }
+
+        for (final Channel channel : channelsToClose) {
+            channel.close().addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) {
+                    channelsToClose.remove(channel);
+                    if (channelsToClose.isEmpty()) {
+                        // last channel was closed - can complete the result promise
+                        // it might already be completed by a concurrently executing listener so use trySuccess
+                        result.trySuccess(null);
```

Comment:
IMHO propagating idle channel closing errors from `SimpleChannelPool.close()` is not very caller-friendly. How would caller handle such error? Would it be useful to propagate an exception with 99 suppressed exceptions if pool had 100 idle channels and all of them failed to close?

Apache commons-pool swallows resource close errors when closed. Javadoc is [here](https://github.com/apache/commons-pool/blob/POOL_2_5_0/src/main/java/org/apache/commons/pool2/ObjectPool.java#L174) and impl is [here](https://github.com/apache/commons-pool/blob/POOL_2_5_0/src/main/java/org/apache/commons/pool2/impl/GenericObjectPool.java#L655-L667). HikariCP also just loggs such erros [here](https://github.com/brettwooldridge/HikariCP/blob/HikariCP-3.2.0/src/main/java/com/zaxxer/hikari/pool/HikariPool.java#L430-L447).

Callers that care about channel close errors can always subscribe to close future of every created channel and gather the errors. Logging errors in the pool could be nice to have.

Reply:
Since the Pool creates the channels, it's responsible for cleaning them up.  It would not be a correct abstraction boundary for callers to watch the close future for a Channel they didn't create.